### PR TITLE
[Sprite Lab] Criterion commands for new CSC module lesson 3

### PR DESCRIPTION
**Pull request by mikeharv**
The CSC Game Development module's second lesson is (mostly) ready for validation. To support this, we need to add new commands related to sprite groups, keyboard events, backgrounds and title screens. This work also required updating our "previous" state to track title screen text and backgrounds.

## Links

Validation Notes Doc: https://docs.google.com/document/d/175wWAuwkDDBDxG31Ed57E7bo5gzACGQ9ZcAHSgk54Xs/edit
Jira : https://codedotorg.atlassian.net/browse/CT-406

## Testing story

I tested this by adding temporary validation to several levels in Katie's progression according to the document above. I tested that level success or failure feedback was triggered as expected.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/p5lab/spritelab/commands/criterionCommands.js [hunk 1]
```diff
@@ -375,6 +375,60 @@ export const commands = {
...
+    const groupSpriteIds = spriteIds.filter(
+      id => this.nativeSpriteMap[id].group === group
+    );
+    return groupSpriteIds.length >= min;
+  },
+
+  // Returns true if a minimum number of sprites have no group.
+  minimumNonGroupSprites(min = 1) {
+    const spriteIds = this.getSpriteIdsInUse();
+    const groupSpriteIds = spriteIds.filter(
```
Comment (ebeastlake) [inline apps/src/p5lab/spritelab/commands/criterionCommands.js]:
```
const nonGroupSpriteIds = spriteIds.filter(id => !this.nativeSpriteMap[id].group);
return nonGroupSpriteIds.length >= min;
```

Comment (mikeharv) [reply apps/src/p5lab/spritelab/commands/criterionCommands.js]: Good catch! Copy/paste error

File: apps/src/p5lab/spritelab/commands/criterionCommands.js [hunk 2]
```diff
@@ -375,6 +375,60 @@ export const commands = {
...
+      id => !this.nativeSpriteMap[id].group
+    );
+    return groupSpriteIds.length >= min;
+  },
+
+  // Returns true if sprites with a minimum number of costumes are in the specified group.
+  minimumCostumesForGroup(group, min = 1) {
+    const uniqueCostumes = [];
+    const spriteIds = this.getSpriteIdsInUse();
+    const spritesInUse = spriteIds.map(id => this.nativeSpriteMap[id]);
```
Comment (ebeastlake) [inline apps/src/p5lab/spritelab/commands/criterionCommands.js:400]: Is `this.nativeSpriteMap[id]` an object that's either defined or undefined? (Just double-checking it's not a number)

Comment (ebeastlake) [reply apps/src/p5lab/spritelab/commands/criterionCommands.js:400]: Yeah, looks like it is!

File: apps/src/p5lab/spritelab/commands/criterionCommands.js [hunk 3]
```diff
@@ -375,6 +375,60 @@ export const commands = {
...
+    const groupSprites = spritesInUse.filter(sprite => sprite.group === group);
+
+    for (const sprite of groupSprites) {
+      const costume = sprite.getAnimationLabel();
+      if (!uniqueCostumes.includes(costume)) {
```
Comment (ebeastlake) [inline apps/src/p5lab/spritelab/commands/criterionCommands.js:405]: nit: In the future, could use a set for this, but no need to change.

File: apps/src/p5lab/spritelab/commands/criterionCommands.js (hunk 2 above)
Comment (mikeharv) [reply apps/src/p5lab/spritelab/commands/criterionCommands.js:400]: Yeah, it should be a full sprite object not just an id. There are some mistakes to that effect in the commit history that I believe I fixed.

File: apps/src/p5lab/spritelab/commands/criterionCommands.js [hunk 4]
```diff
@@ -375,6 +375,60 @@ export const commands = {
...
+    return groupSpriteIds.length === 1;
+  },
+
+  // Returns true if a player sprite has an upward vertical velocity
+  playerSpriteJumping() {
+    const spriteIds = this.getSpriteIdsInUse();
+    const spritesInUse = spriteIds.map(id => this.nativeSpriteMap[id]);
+
+    const jumpingSprites = spritesInUse.filter(
+      sprite => sprite.group === 'players' && sprite.velocityY < 0
```
Comment (ebeastlake) [inline apps/src/p5lab/spritelab/commands/criterionCommands.js:427]: Is this supposed to be `< 0` and not `> 0`?

Comment (mikeharv) [reply apps/src/p5lab/spritelab/commands/criterionCommands.js:427]: No it’s correct as is. The velocity is negative if the sprite is moving up. This based on the actual P5 orientation and not a Sprite Lab block that remaps it in the other direction.

File: apps/src/p5lab/spritelab/commands/criterionCommands.js (hunk 3 above)
Comment (mikeharv) [reply apps/src/p5lab/spritelab/commands/criterionCommands.js:405]: I had that thought but wasn’t which was preferred/more readable.

File: apps/src/p5lab/spritelab/commands/criterionCommands.js [hunk 5]
```diff
@@ -512,6 +586,42 @@ export const commands = {
...
+    if (this.previous.background) {
+      return this.getBackground() !== this.previous.background;
+    } else {
+      return false;
+    }
+  },
+
+  // Returns true if the title text changed this frame.
+  titleChanged() {
+    const previousScreenText = this.previous.screenText;
```
Comment (ebeastlake) [inline apps/src/p5lab/spritelab/commands/criterionCommands.js:614]: Here and above, is `this.previous` guaranteed to exist?

File: apps/src/p5lab/spritelab/commands/criterionCommands.js (hunk 3 above)
Comment (ebeastlake) [reply apps/src/p5lab/spritelab/commands/criterionCommands.js:405]: Either is fine! There aren't any performance issues with an array this size, so it really doesn't matter.

File: apps/src/p5lab/spritelab/commands/validationCommands.js [hunk 6]
```diff
@@ -223,6 +223,8 @@ export const commands = {
     this.previous.printLogLength = this.printLog.length || 0;
     this.previous.soundLogLength = this.soundLog.length || 0;
     this.previous.foregroundEffectsLength = this.foregroundEffects.length || 0;
+    this.previous.background = this.background;
```
Comment (ebeastlake) [inline apps/src/p5lab/spritelab/commands/validationCommands.js:226]: I think this answers my previous question.

**Review by ebeastlake — APPROVED** [review]

File: apps/src/p5lab/spritelab/commands/criterionCommands.js (hunk 5 above)
Comment (mikeharv) [reply apps/src/p5lab/spritelab/commands/criterionCommands.js:614]: Yes it is!

**Pull request merged**